Bitsandbytes precision plugin (#18655)
Lightning-AI/lightning: docs/source-pytorch/common/precision_intermediate.rst

Both 4-bit (`paper reference <https://arxiv.org/abs/2305.14314v1>`__) and 8-bit (`paper reference <https://arxiv.org/abs/2110.02861>`__) quantization is supported.
Specifically, we support the following modes:
* nf4: Uses the normalized float 4-bit data type. This is recommended over "fp4" based on the paper's experimental results and theoretical analysis.
* nf4-dq: "dq" stands for "Double Quantization" which reduces the average memory footprint by quantizing the quantization constants. In average, this amounts to about 0.37 bits per parameter (approximately 3 GB for a 65B model).
* fp4: Uses regular float 4-bit data type.
* fp4-dq: "dq" stands for "Double Quantization" which reduces the average memory footprint by quantizing the quantization constants. In average, this amounts to about 0.37 bits per parameter (approximately 3 GB for a 65B model).
* int8: Uses unsigned int8 data type.
* int8-training: Meant for int8 activations with fp16 precision weights.

While these techniques store weights in 4 or 8 bit, the computation still happens in 16 or 32-bit (float16, bfloat16, float32).
This is configurable via the dtype argument in the plugin.

> Quantizing the model will dramatically reduce the weight's memory requirements but might impact performance negatively.

Reviewer: Here, what you mean is performance in terms of predictive performance (e.g. accuracy of the classifier), not runtime performance. This could be clarified.

> Quantizing the model will dramatically reduce the weight's memory requirements but might impact performance negatively.

Reviewer: ```suggestion
Quantizing the model will dramatically reduce the weight's memory requirements but might impact predictive performance negatively.
```

> Quantizing the model will dramatically reduce the weight's memory requirements but might impact performance negatively.

Reviewer: ```suggestion
Quantizing the model will dramatically reduce the weight's memory requirements but might impact the accuracy of your model negatively.
```

The Trainer automatically replaces the :class:`torch.nn.Linear` layers in your model with their BNB alternatives.

.. code-block:: python

    from lightning.pytorch.plugins import BitsandbytesPrecisionPlugin

    # this will pick out the compute dtype automatically, by default `bfloat16`
    precision = BitsandbytesPrecisionPlugin("nf4-dq")
    trainer = Trainer(plugins=precision)

    # Customize the dtype, or skip some modules

>     precision = BitsandbytesPrecisionPlugin("int8-training", dtype=torch.float16, skips={"lm_head"})

Reviewer: Typically arguments that describe an action are in the imperative form. "skips" could both be a verb or a noun. A better choice might be ignore or exclude. Options that I like more:

- ignore
- exclude
- ignore_modules 
- exclude _modules
- skip
- skip_modules

The `_modules` version would make it very clear we are expecting module names not parameter names.

    trainer = Trainer(plugins=precision)

    class MyModel(LightningModule):
        def configure_model(self):
            # instantiate your model in this hook
            self.model = MyModel()

.. note::

    Only supports CUDA devices and the Linux operating system. Windows users should use
    `WSL2 <https://learn.microsoft.com/en-us/windows/ai/directml/gpu-cuda-in-wsl>`__.

> This plugin does not take care of replacing your optimizer with an 8-bit optimizer e.g. `bitsandbytes.optim.Adam8bit``.

Reviewer: ```suggestion
This plugin does not take care of replacing your optimizer with an 8-bit optimizer e.g. ``bitsandbytes.optim.Adam8bit``.
```
unbalanced ticks
